reconcile/statuspage/atlassian.py

```diff
@@ -30,126 +35,316 @@ class AtlassianComponent(BaseModel):
         else:
-            LOG.info(f"create component {desired.name}: {component_update}")
+            logging.info(f"create component {desired.name}: {component_update}")
             if not dry_run:
-                return self._create_component(component_update)
-            else:
-                return None
-
-    def _update_component(self, id: str, data: dict[str, Any]) -> None:
-        self._client().components.update(id, **data)
-
-    def _create_component(self, data: dict[str, Any]) -> Optional[str]:
-        result = self._client().components.create(**data)
-        return result.get("id")
-
-    def delete_component(self, dry_run: bool, id: str) -> None:
-        if not dry_run:
-            self._client().components.delete(id)
-            self.rebuild_state()
+                component_id = self._api.create_component(component_update)
+                self._bind_component(
+                    dry_run=dry_run,
+                    component_name=desired.name,
+                    component_id=component_id,
+                )
 
-    def _fetch_components(self) -> list[AtlassianComponent]:
-        raw_components = self._client().components.list()
-        group_ids_to_name = {g.id: g.name for g in raw_components if g.group}
-        return [
-            AtlassianComponent(
-                **c.toDict(), group_name=group_ids_to_name.get(c.group_id, None)
+    def delete_component(self, dry_run: bool, component_name: str) -> None:
+        component_id = self._binding_state.get_id_for_component_name(component_name)
+        if component_id:
+            if not dry_run:
+                self._api.delete_component(component_id)
+                self._binding_state.forget_component(component_name)
+                self._build_component_cache()
+        else:
+            logging.warning(
+                f"can't delete component {component_name} because it is not "
+                f"bound to any component on page {self.page_name}"
             )
-            for c in raw_components
-        ]
 
-    def _client(self):
-        return statuspageio.Client(
-            api_key=self.token, page_id=self.page_id, organization_id="unset"
+    def has_component_binding_for(self, component_name: str) -> bool:
+        return self._binding_state.get_id_for_component_name(component_name) is not None
+
+    def _bind_component(
+        self,
+        dry_run: bool,
+        component_name: str,
+        component_id: str,
+    ) -> None:
+        logging.info(
+            f"bind component {component_name} to ID {component_id} "
+            f"on page {self.page_name}"
         )
+        if not dry_run:
+            self._binding_state.bind_component(component_name, component_id)
 
 
-def load_provider(page: StatusPage) -> StatusPageProvider:
-    provider = AtlassianStatusPage(
-        page_id=page.page_id,
-        api_url=page.api_url,
-        token=page.credentials.get("token"),
+def init_provider_for_page(
```

Comment:
TBH, I don't like the coupling to `page` and `secret_reader` here (ok maybe `page` can be argued somehow). What do you think about:

```python
def init_provider_for_page(page_id, name, api_url, token, component_binding_state)
```

Reply:
i agree about the secret_reader. i will replace it to make the provider ignorant about the source of the token, which is a good thing. i will keep the page object for now though.